paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc

```diff
@@ -0,0 +1,421 @@
+  graph_ = graph;
+  BuildOpDependencyMap();
+  for (size_t i = 0; i < ScopeNum(); ++i) {
+    RunOnScopeIdx(i);
+  }
+}
+
+// Note(zjl): The reason why I separate SortOp from BuildOpDependencyMap()
+// is that we can use different sorting strategies in the future to
+// evaluate the effects of different sorting strategies.
+// Currently, I use BFS, but we can use other kinds of sorting strategy
+// in the future, as long as the new strategy reaches higher memory reuse
+// ratio.
+std::vector<OpHandleBase *> BufferSharedCrossOpMemoryReusePass::SortOp(
+    const OpGraphView &graph_view) const {
+  std::vector<OpHandleBase *> sorted_ops;
+  sorted_ops.reserve(graph_view.OpNumber());
+  graph_view.BreadthFirstVisit(
+      [&](OpHandleBase *cur_op) { sorted_ops.emplace_back(cur_op); });
+  PADDLE_ENFORCE_EQ(sorted_ops.size(), graph_view.OpNumber(),
+                    "There are unvisited ops");
+  return sorted_ops;
+}
+
+/**
+ * Try to reuse unlived vars.
+ *
+ * What we do is: transverse all outputs of each op, and find a suitable
+ * unused var, and then reuse its memory as output.
+ *
+ * How to determine unused vars?
+ *
+ * Case 1: unlived vars after all preceding ops run. In this case, no extra
+ *   edge would be added to the graph.
+ *
+ * Case 2: unlived vars after all preceding ops and all no-dep ops run. In
+ *   this case, the reused var is from no-dep ops, so that we have to add
+ *   extra edge to resolve data hazard.
+ *
+ *
+ * If Case 2 occurs, what we should do to resolve data hazard?
+ *
+ *  - Step 1: add a dep var between reused_op and share_tensor_buffer_op,
+ *            that is: reused_op -> dep_var -> share_tensor_buffer_op.
+ *
+ *  - Step 2: Update deps_, all preceding ops of reused_op should be
+ *            preceding ops of op.
+ */
+void BufferSharedCrossOpMemoryReusePass::RunOnScopeIdx(size_t idx) const {
+  auto &ops = ops_[idx];
+
+  auto &last_live_ops_of_vars =
+      Get<std::vector<LastLiveOpsOfVars>>(kLastLiveOpsOfVars)[idx];
+
+  // Build a reverse map of `last_live_ops_of_vars`,
+  // i.e., VarHandle -> last lived ops of VarHandle
+  std::unordered_map<VarHandle *, std::unordered_set<ComputationOpHandle *>>
+      var_to_ops;
+  for (auto &pair : last_live_ops_of_vars) {
+    for (auto *op : pair.second.ops()) {
+      var_to_ops[pair.second.var()].insert(op);
+    }
+  }
+
+  // Deep copy of `var_to_ops`, used to get last lived ops of each unlived var
+  auto original_var_to_ops = var_to_ops;
+
+  // Memory size of VarHandle -> list<VarHandle>
+  std::map<int64_t, std::list<VarHandle *>> unlived_var_pool;
+  size_t reuse_num = 0;
+
+  for (auto *op : ops) {
+    // Transverse all output args of op, find whether there is unlived var
+    // can be reused.
+    auto out_args = op->Node()->Op()->OutputArgumentNames();
+    for (auto &out_arg : out_args) {
+      auto out_nodes = this->FindNodesByName(out_arg, op->Node()->outputs);
+      if (out_nodes.size() != 1) {
+        continue;
+      }
```

Comment:
Why the out_nodes may be empty?

Reply:
It may happen when `out_arg == kEmptyVarName`.